---
title: "ðŸ“Š HT2024: Data-driven Life Sciences"
linkTitle: Overview
summary: Course content of the Data-driven Life Sciences course offered in 2024.
date: '2024-05-01'
type: book
tags:
  - current
---

{{< figure src="featured.jpg" >}}

{{< toc hide_on="xl" >}}



Welcome to the Data-driven Life Sciences 2024 course, where you will explore the intersection of data science, artificial intelligence, and life sciences to drive innovation and discovery. This fully online course culminates in an in-person hackathon, fostering a vibrant community that gathers the DDLS and SciLifeLab members.

The 6 modules aim to introduce learners to computer-driven life sciences, covering application areas in data-driven life sciences. Guest lecturers (DDLS Fellows, SciLifeLab fellows, and SciLifeLab facility training providers) will teach topics including technologies and analysis of data sets from proteomics, transcriptomics, biomolecular structure, molecular dynamics simulations, and various imaging techniques. These modules present, analyze, and discuss models of biological phenomena and related scientific breakthroughs based on such data analysis.

The course is open to **all the master's, PhD students and anyone who interested in Sweden**, please use the following link to register:

 {{< cta cta_text="Register Now" cta_link="https://forms.gle/9gmxpEE6X6HZ1wLa7" >}}

You can either register for all the modules or individual modules.

Specifically students at KTH, you should register for the course at the KTH course selection system:
 - [`SK2538` (master's)](https://www.kth.se/student/kurser/kurs/SK2538)
 - [`FSK3538` (PhD)](https://www.kth.se/student/kurser/kurs/FSK3538): PhD student sign up [here](https://forms.gle/9gmxpEE6X6HZ1wLa7).


## Intended Learning Outcomes

-   Describe the field of data-driven life sciences

-   Present an overview of various application areas

-   Provide examples of applications and their associated analysis methods

-   Apply statistical and machine learning analysis to biological data sets

-   Formulate models of biological phenomena

-   Present and review scientific literature in computer-driven life sciences

-   Reflect on the ethical consequences of data-driven life sciences

-   Practice good data management, including collection, handling, sharing, and analysis

## Course Format and Credits

The course consists of 6 modules: a basic module and 5 elective modules. Each module is a stand-alone short-form course with 3 sessions per module. Completing all 6 modules earns a total of 7.5 ECTS.

## Modules and Certification
Each course module contains 3 main activities: a lecture on Tuesday, giving background information about the topic(s) of the week; a computer lab, usually on Wednesday, where you will practically explore data exploration using Python in Jupyter notebooks; and a journal club on Friday where you will collectively explain and discuss an assigned research paper.

Both the computer lab and the journal club are mandatory and graded activities!

- For the labs, you be given a Jupyter notebook with tutorials and coding exercises. You will be running and practicing coding by running the Jupyter notebook in [Google Colab](https://colab.research.google.com/?utm_source=scs-index). During the lab session, you can discuss your answers with the lab teachers. You will submit your final notebook with answers to the questions to be graded.

- For the journal club, you will have a list of questions (the same every week) which we will answer during the journal club. Your participation and answer to these questions will be graded during the journal club.

In addition, in the final project week / hackathon, course participants will decide on a project to be conducted in pairs (Ms students) or alone (PhD students) and followed by an oral presentation. The project and presentation will be graded by peers and the examiner.


You can sign up partially or for the whole course. Each module is worth a certain number of ECTS credits, and you can choose to participate in the modules that interest you most. However, to obtain a certification for participation, you must complete all modules.

Module 1 (basic module) is a prerequisite for obtaining certification for attendance in any of the subsequent elective modules. To receive a certificate, you must:

-   Apply for and participate in Module 1.

-   Actively participate in all registered modules by attending and engaging in all sessions.

You will receive 0.5 ECTS for the basic module and 1 ECTS for each elective module. Completing a final project earns an additional 2 ECTS. For example, if you are a PhD student interested in Module 2, you must complete both Module 1 and Module 2 to receive 0.5 + 1 ECTS. Master's students from KTH or other universities must complete the entire course, including the basic and elective modules, plus the final project.

The first module covers course basics and prerequisites. The subsequent modules each run for a week and cover various topics in data-driven life sciences (DDLS). Each week includes three main activities:

-   One or two 2-hour lectures on Tuesday providing background information about the week's topic(s)

-   A computer lab on Wednesday for practical data exploration using Python in Jupyter notebooks (Google Colab)

-   A seminar on Friday for collective explanation and discussion of an assigned research paper


 {{< cta cta_text="Register Now" cta_link="https://forms.gle/9gmxpEE6X6HZ1wLa7" >}}


You can either register for the full course or selected modules.

## Self-directed Learning

One of the principal objectives of this course is to instill in you a data-driven mindset essential for life science research. More importantly, we aim to teach you the art of "learning how to learn." This involves honing your ability to independently acquire new knowledge and skills. ChatGPT serves as an excellent tool to facilitate this self-directed learning process. Instead of focusing solely on the course content, we encourage you to embrace this broader, skill-centric approach to education. We will use this self-directed learning approach extensively in the computer lab and journal club sessions.


## Course Modules

**Note: The exact content for each module may be subject to change. Please refer to the updated schedule for the most current information.**

### Module 1 [0.5 ECTS] - Introduction to DDLS, AI and ChatGPT

**NOTE: Module 1 is a prerequisite for any subsequent Module if you like to obtain a certification for participation.**

**Start date**: Week 35 (August 27th, 2024)

#### Speaker: Wei Ouyang (DDLS Fellow, 8:00-10:00 CEST)

Module 1 serves as the foundational introduction to the Data-Driven Life Sciences (DDLS) course. This basic module is designed to familiarize participants with the overarching principles and goals of DDLS, getting prepared for machine learning topics and providing a comprehensive overview of the course structure, content, and logistics. Participants will be introduced to the various topics that will be covered in the subsequent modules, as well as the format of the course, which includes lectures, computer labs, and seminars. The module will also outline the examination and certification requirements.

A key focus of this module will be on introducing tools and tips to facilitate self-learning, including the basics of machine learning, running and using Google Colab, and prompt engineering in ChatGPT. Participants will learn how to use ChatGPT to enhance their learning experience by digesting course material, understanding new concepts, and assisting with coding tasks in the computer lab.

By the end of this module, attendees will have a solid understanding of what to expect in the course, including the practical and theoretical components, ensuring they are well-prepared for the advanced topics in the following modules.

### Module 2 [1 ECTS] - Advanced AI Techniques in Biological Systems

**Start date**: Week 36 (September 3rd, 2024)

#### Speaker 1: Patrick Bryant (DDLS Fellow, 8:00-9:00 CEST)

This module will explore AI applications in biology, including techniques for predicting protein structures and interactions using AlphaFold2. The session will also cover how reinforcement learning and deep learning can be used to design new molecules for biotechnological applications.

#### Speaker 2: Johan Bengtsson-Palme (DDLS Fellow, 9:00-10:00 CEST)

This session will delve into microbiology, focusing on antibiotic resistance through bioinformatics and metagenomics. It will discuss the use of large-scale DNA sequencing to study microbial communities and antibiotic resistance, and explore methods for accurate taxonomic determinations based on sequence data.

**Computer Lab**: You will learn how to use AlphaFold2 to predict protein structures through hands-on exercises, gaining practical experience in applying advanced AI techniques to biological systems.

### Module 3 [1 ECTS] - Bioinformatics and Metagenomics

**Start date**: Week 37 (September 10th, 2024)

#### Speaker 1: Nima Rafati (Data Scientist at NBIS, 9:00-10:00 CEST)

This module will explore bioinformatics, focusing on the use of next-generation sequencing (NGS) to uncover the genetic basis of complex traits across various species. Topics will include genome assembly, single-cell sequencing, and the development of genetic tests.

#### Speaker 2: Laura Carroll (DDLS Felow, September 11th, 2024; 10:00-11:00 CEST)

This session will discuss the impact of (meta)genomic sequencing in microbiology. It will cover bioinformatic approaches for pathogen surveillance and outbreak detection, and the development of phylodynamic models and machine learning techniques for microbial analysis.

**Computer Lab**: The lab will cover tools in bioinformatics and metagenomics data analysis with practical exercises, providing hands-on experience in applying these techniques to real-world datasets.

### Module 4 [1 ECTS] - AI in Bioimage Analysis

**Start date**: Week 38 (September 17th, 2024)

#### Speaker 1: Wei Ouyang (DDLS Fellow, 8:00-9:00 CEST)

This module will introduce the basic concepts in recent advances in generative AI, large language models, and diffusion models, focusing on creating conditional microscopy cell images for whole-cell modeling. It will also cover challenges in generative AI and potential paths forward in data-driven life sciences.

#### Speaker 2: Estibaliz GÃ³mez de Mariscal (Postdoc at Instituto Gulbenkian de CiÃªncia, Portugal, 9:00-10:00 CEST)

This session will explore bioimage analysis, highlighting the application of deep learning to microscopy images for biological insights. Topics will include deep learning models for image analysis, best practices for training and performing deep learning methods for microscopy image analysis, and tools and programming considerations for applying these techniques to real datasets.

**Computer Lab**: The lab will provide hands-on experience with tools and techniques in bioimage analysis, including the use of deep learning models for processing and analyzing microscopy images.

### Module 5 [1 ECTS] - Single Molecule Analysis and Tissue Atlases

**Start date**: Week 39 (September 24th, 2024)

#### Speaker: Daniel FÃ¼rth (SciLifeLab Fellow, 8:00-10:00 CEST)

This module will explore advanced tools for single-molecule imaging and programmable bioconjugation to study cell communication. It will cover essential computational methods for image analysis of single molecules in both fixed and living cells. Participants will learn about the segmentation of individual molecules in 2D and 3D using U-Net and ResNet deep learning architectures, and how to track these molecules over time using DeepSORT, an enhanced version of the SORT algorithm.

The session will also discuss transforming images of tissue sections onto a known anatomical reference atlas using rigid and non-rigid transformations. Participants will explore point-cloud registration algorithms like Thin-Plate Splint (TPS) registration and Coherent-Point Drift (CPD), comparing their relative merits. Additionally, the lecture will introduce valuable resources, database structures, and ontologies relevant to creating anatomical atlases.

**Computer Lab**: The lab session will involve hands-on exercises using Jupyter notebooks to apply the discussed computational tools for analyzing fluorescent microscopy data. Participants will practice segmenting and tracking single molecules, as well as performing tissue atlas registration.

### Module 6 [1 ECTS] - Computational Modeling and Automated Scientific Discovery

**Start date**: Week 40 (October 1st, 2024)

#### Speaker 1: Cemal Erdem (DDLS Fellow, 8:00-9:00 CEST)

This module will focus on combining machine learning, mechanistic models, and patient data to develop predictive computational models for cancer and other diseases. Topics will include mathematical modeling of signaling cascades, systems biology, quantitative systems pharmacology (QSP), and precision medicine.

#### Speaker 2: Gabriel Reder (PostDoc at KTH & Cambridge University, 9:00-10:00 CEST)

This session will cover the integration of computer science, math, and software engineering to automate laboratory processes. Topics will include lab robotics, ion mobility mass spectrometry, and lab automation. Participants will learn about a multi-agent collaboration system capable of hypothesis generation, experimental design, protocol compilation, and data analysis, aiming to create autonomous systems for automatic scientific discovery.

**Computer Lab**: The lab will feature examples of cell modeling and using GPT models to enable automation. Participants will gain hands-on experience with computational modeling techniques and explore how large language models can be utilized to streamline and automate laboratory processes. This session will provide practical insights into the integration of AI tools for enhancing research efficiency and innovation.

### Final Project: Hackathon at SciLifeLab - for course participants and DDLS fellows

**Start date**: Week 41 (October 9th, 2024)

Join us at the DDLS Hackathon at SciLifeLab in Solna, where course participants will choose and collaborate on a range of projects provided by us. With the support of teaching assistants in a hands-on environment, groups will innovate and develop solutions in the dynamic field of data-driven life science.

The final week is a project week where participants will select a dataset and decide how to apply the skills and knowledge gained during the course to create AI models, analyze the data, and produce meaningful insights. To make it more interactive, all participants are invited to come to SciLifeLab in Solna for a week-long hackathon. At the end of the week, there will be concluding presentations for the teams to show off what they achieved during the hackathon.

Everyone involved in the DDLS program, including PIs, affiliated PhD students, postdocs, and facility trainers, are welcome to join the code fair. Participants will form pairs or small groups of 2-3 people on-site to work on projects, exchange ideas and experiences, and engage in pair programming. This collaborative environment will foster innovation and provide a unique opportunity to apply the concepts learned throughout the course to real-world problems in data-driven life sciences.

## Grades

The final grade on the course is determined as follows:

-   The computer labs are graded P/F. To pass, you need to attend all the labs and answer the questions to a satisfactory degree.
-   Participation in the seminars is graded as P/F. To pass, you need to read the assigned papers, attend all the seminars and participate in discussions.
-   The project is graded P/F. To pass, you need to carry out the project, and get a passing grade on the project report.
-   The oral exam (project presentation) at the end of course will be graded A-F scale.

-   For master students, the grade on the oral exam determines the final grade on the course provided the other three activities have received a passing grade.
-   For PhD students, the oral exam is not mandatory, but you are welcome to join the oral exam if you want to practice your presentation skills.

The submission of the computer lab, the project report and the oral exam will be first reviewed and graded by your peers, and then the final grade and feedbacks will be given by the teachers and examiner.

## Communication and groups

You can find all announcements at [here](/post/).

For questions, please email the course responsible, see contact [here](/contact/).

## Meet your instructor

In this course, you will meet:
-   {{< mention "wei" >}}, assistant professor in biophysics who will be holding the lectures, seminars and grading them (<weio@kth.se>). Wei is also the course responsible.
-   {{< mention "songtao" >}}, PhD student in biophysics, who will be teaching assistant for the course, chairing lectures, holding the computer labs and journal clubs [songtao.cheng@scilifelab.se)](mailto:songtao.cheng@scilifelab.se)
-   {{< mention "nils" >}}, PhD student in biophysics, who will be teaching assistant for the course, chairing lectures, holding the computer labs and journal clubs(<nils.mechtel@scilifelab.se>)


## Course schedule

See the [course schedule](./schedule/) for the detailed schedule.

## FAQs

{{< spoiler text="Are there prerequisites?" >}}
The students are expected to have basic knowledge of biology and programming in Python. If you are not familiar with Python, it will be helpful if you can go through the [prerequisites](./prerequisites).
{{< /spoiler >}}

{{< spoiler text="How often do the courses run?" >}}
We run the course once a year, in Period 1 from August to October.
{{< /spoiler >}}

{{< spoiler text="How do I register the course?" >}}
-   For KTH master students, please register at the KTH course selection system.
-   For PhD students or anyone from other universities, please sign up at [here](https://forms.gle/9gmxpEE6X6HZ1wLa7).
-   You can also sign up for individual module(s) [here](https://forms.gle/9gmxpEE6X6HZ1wLa7). 
{{< /spoiler >}}

{{< spoiler text="How do I get access to the course material?" >}}
The course material are available at [here](https://ddls.aicell.io). We won't be using KTH Canvas.
{{< /spoiler >}}

{{< spoiler text="Can I attend the course remotely?" >}}

It is possible only for the lectures. Except the first lecture (in person), all other lectures will be given over Zoom (zoom link: https://kth-se.zoom.us/j/69812177998).

The computer labs and seminars are provided in person. Exceptions will be announced in the email and the updated schedule above.
{{< /spoiler >}}

{{< spoiler text="How do I get access to the computer labs?" >}}
The computer labs are normally provided in person (exception may apply, check email announcement and the updated schedule). If you are not able to attend, please contact the course responsible.
{{< /spoiler >}}

{{< spoiler text="How do I get access to the seminars?" >}}
The seminars are normally provided in person (exception may apply, check email announcement and the updated schedule). If you are not able to attend, please contact the course responsible.
{{< /spoiler >}}

{{< spoiler text="What if I cannot attend the computer labs or seminars?" >}}
It is mandatory to attend all the computer labs or seminars, if you anticipate that you will miss a computer lab or seminar, please notify the course responsible as soon as possible, exceptions will be made on a case-by-case basis.

If the exception is granted, you will still need to submit the notebook for the computer lab, or write a report to answer the seminar question sheet.

{{< /spoiler >}}

{{< spoiler text="Will the lecture be recorded?" >}}
The lecture will not be recorded, but the slides will be made available after the lecture.
{{< /spoiler >}}

{{< spoiler text="Can I use ChatGPT / Claude or other generative AI tools in the computer labs, seminar, assignments or the final project?" >}}
Yes, you are encouraged to use these AI tools to facilitate your learning, and perform tasks. However, you need to make sure that you are using them in a responsible way. Importantly, if you used it for submitting any graded activities, please make sure to also attach the exported conversation history with the AI tool.
{{< /spoiler >}}

{{< cta cta_text="Begin the course" cta_link="module-1" >}}
