---
title: "Module 2 [1 ECTS] - Advanced AI Techniques in Biological Systems"
summary: " "
linkTitle: "Module 2"
date: '2024-06-13'
type: book
weight: 30
course_module: "true"
---
<style>
  .profile-photo {
    width: 150px; /* Adjust the width as needed */
    height: auto; /* This keeps the aspect ratio of the image */
    display: block;
    margin-left: auto;
    margin-right: auto;
  }
</style>

**Start date**: Week 36 (September 3rd, 2024)

<!-- has been updated -->
#### Speaker 1: Patrick Bryant (Ass. Prof. at SU & DDLS Fellow, 8:00-9:00 CEST)

My research tries to answer questions about the evolution of proteins and how this information can be used to create a new range of AI tools. [Google Scholar]{https://scholar.google.com/citations?user=KPlaFQQAAAAJ&hl=sv&oi=ao}

**Title: Protein structure prediction and design**

The foundation of protein design is accurate structure prediction which depends on coevolutionary patterns found in multiple sequence alignments (MSAs). Coevolution constrains the protein structure just enough for it to become predictable given a suitable MSA. However, designing new proteins poses a challenge in the absence of coevolutionary data. Therefore, protein design methodologies must incorporate learned associations between amino acid sequences and protein structures, encapsulating the rules governing the known protein space. In this lecture, you will learn more about how to design new proteins and in the practical session you will design your own peptide binder using the latest structure prediction technology.

<!-- has been updated -->
#### Speaker 2: Johan Bengtsson-Palme (DDLS Fellow at Chalmers University of Technology, 9:00-10:00 CEST)

{{< figure src="Johan_Bengtsson-Palme.jpg" class="profile-photo"  >}}

I have a PhD in Medicine from the University of Gothenburg focused on the role of the environment in antibiotic resistance development and did a postdoc on interactions in microbial communities with Prof. Jo Handelsman at the University of Wisconsin-Madison (USA). In May 2022, where I lead a research group working on data-driven microbiology. Read more on the group website at [https://microbiology.se]{https://microbiology.se}

**Title: Large-scale predictions of antibiotic resistance in microbial communities**

This lecture will focus on how we can predict antibiotic resistance and virulence in bacterial communities through large-scale experiments, metagenomics and bioinformatics. We will discuss methods to identify antibiotic resistance genes, investigate their potential for transfer to pathogens, and predict where bacteria carrying resistance are found. We will integrate methods under development in the EMBARK and SEARCHER programs ([http://antimicrobialresistance.eu]{http://antimicrobialresistance.eu}).

<!-- needs to be updated -->
**Computer Lab**: You will learn how to use AlphaFold2 to predict protein structures through hands-on exercises, gaining practical experience in applying advanced AI techniques to biological systems.