---
title: "Module 3 [1 ECTS] - Bioinformatics and Metagenomics"
summary: ""
linkTitle: "Module 3"
date: "2024-06-13"
type: book
weight: 40
course_module: "true"
---
<style>
  .profile-photo {
    width: 150px; /* Adjust the width as needed */
    height: auto; /* This keeps the aspect ratio of the image */
    display: block;
    margin-left: auto;
    margin-right: auto;
  }
</style>

**Start date**: Week 37 (10th September 2024)

**Note**: This module contains two lectures held on two different dates: September 10th Tuesday at 9:00-10:00 CEST and September 11th Wednesday at 10:00-11:00 CEST.

<!-- updated: 31.08.2024 -->
#### Speaker 1: Nima Rafati (Data Scientist at NBIS, SciLifeLab, Uppsala University; Tuesday 10th September 9:00-10:00 CEST)

{{< figure src="Nima_Rafati.jpg" class="profile-photo">}}

Nima Rafati is a passionate bioinformatician focused on uncovering patterns in biological data through the analysis of sequencing data across various living systems.

**Title: Decoding Cell-Level Biology with scRNA-Seq Data**

The lecture will cover the scRNA-Seq workflow, including sample preparation, sequencing, data preprocessing, and result interpretation. Key topics such as quality control and clustering will be discussed. An example of scRNA-Seq data analysis will be demonstrated to illustrate these concepts.

<!-- updated: 28.06.2024 -->
#### Speaker 2: Laura Carroll (DDLS Fellow; Wednesday 11th September 10:00-11:00 CEST)

{{< figure src="Laura_Carroll.jpeg" class="profile-photo">}}

Dr. Laura Carroll is an Assistant Professor and DDLS Fellow in the Department of Clinical Microbiology at Ume√• University. As a microbial bioinformatician, her research focuses on developing bioinformatic approaches to leverage whole-genome sequencing (WGS) data for improving pathogen surveillance, source tracking, outbreak detection, and risk evaluation. Learn more about her work on her [institutional page](https://www.umu.se/en/staff/laura-carroll/) or visit her [research group website](https://www.microbe.dev/). Find her publications on [Google Scholar](https://scholar.google.com/citations?user=eJHtnUAAAAAJ&hl=en).

**Title: (Meta)genome mining approaches for microbial secondary metabolite discovery**

Microbes can produce a wide range of secondary metabolites, which allow them to interact with each other, as well as with their environment more broadly. Beyond their ecological roles, some microbial secondary metabolites have found uses in medicine or industry (e.g., as first-in-class drugs), while others can have negative impacts on human health (e.g., as toxins or carcinogens). To discover novel microbial secondary metabolites, (meta)genome mining approaches make use of the fact that the enzymatic machinery responsible for secondary metabolite production is encoded by physically clustered groups of genes in microbial genomes, called biosynthetic gene clusters (BGCs). Here, we will discuss state-of-the-art (meta)genome mining approaches for BGC discovery, with a focus on machine learning (ML) approaches. We'll walk through how to train and evaluate a sequence segmentation model for BGC detection tasks, step-by-step, and we'll discuss common ML pitfalls. Finally, we'll discuss how these methods can be applied to (meta)genomic data to gain insight into microbe-microbe and microbe-host interactions.

<!-- TODO: needs to be updated -->
**Computer Lab**: The lab will cover tools in bioinformatics and metagenomics data analysis with practical exercises, providing hands-on experience in applying these techniques to real-world datasets.
